### PR TITLE
feat(graph, graphql, store): Support new `Timestamp` scalar

### PR DESCRIPTION
**Pull request by dotansimha**
## TODO
- [x] Add new scalar
- [x] Handle SQL
- [x] Decide on how to format the values
- [x] PR in graph-tooling (https://github.com/graphprotocol/graph-tooling/pull/1522)
- [x] Make it compile 
- [x] Fix CI and tests
- [x] Use `i64` instead of `string` 
- [x] Add integration tests with `graph-tooling`

### CODE REVIEW

**Comment by lutter** [comment]
I am very interested in getting this merged! Would be great to discuss how to get this PR ready - more than happy to help.

I fixed some of this up and put it onto the branch `lutter/ts2` - it compiles, but there are lots of test failures.

I am also not sure how best to expose timestamps to AS; the way I did it right now is as seconds since the epoch, but that ties AS to second granularity. That might be a little too coarse in the future. We could do microseconds, but we'd also need some support to deal with timestamps in `graph-tooling`, and for that it might be better to expose the timestamps not as an i64 but in a more structured way.

As I said: let's discuss.

**Comment by lutter** [comment]
Thinking about this some more, we should probably expose it in microseconds - an `i64` would cover roughly +/- 290k years in microseconds which I think is plenty of range for timestamps.

**Comment by dotansimha** [comment]
> Thinking about this some more, we should probably expose it in microseconds - an `i64` would cover roughly +/- 290k years in microseconds which I think is plenty of range for timestamps.

Yeah, this makes sense. Thank you @lutter for the help. I'm back on this today, focused on graph-tooling PR.
Will ping you when I have some progress!

File: store/postgres/src/relational_queries.rs [hunk 1]
```diff
@@ -362,6 +367,14 @@ impl FromColumnValue for r::Value {
         }
     }
 
+    fn from_timestamp(i: &str) -> Result<Self, StoreError> {
+        scalar::Timestamp::from_rfc3339(i)
+            .map(|v| r::Value::Int(v.as_millis_since_epoch()))
```
Comment (lutter) [inline store/postgres/src/relational_queries.rs:372]: If we had a `r::Value::Timestamp` we could avoid this conversion. And isn't that going to run into integer overflow at some point>

File: store/test-store/tests/graphql/query.rs [hunk 2]
```diff
@@ -459,12 +461,13 @@ async fn insert_test_entities(
     let is = &manifest.schema;
     let pub1 = &*PUB1;
     let ts0 = BlockTime::for_test(&BLOCKS[0]);
+    let timestamp = Timestamp::from_millisecs_since_epoch(679736164000);
```
Comment (lutter) [inline store/test-store/tests/graphql/query.rs:464]: Happy birthday!

**Review by lutter — CHANGES_REQUESTED** [review]
Awesome! This looks good; the two main things I am not sure about are:

- passing timestamps between mappings and graph-node as milliseconds. Should we use microseconds to make sure we never have to change how this works?
- converting inputs to internal timestamp types earlier rather than passing strings around (`r::Value::Timestamp`)

File: graph/src/data/graphql/object_macro.rs [hunk 3]
```diff
@@ -52,6 +52,13 @@ impl IntoValue for u64 {
     }
 }
 
+impl IntoValue for i64 {
+    #[inline]
+    fn into_value(self) -> r::Value {
+        r::Value::String(self.to_string())
+    }
+}
```
[the remark this answers is not in the document]
Comment (dotansimha) [reply graph/src/data/graphql/object_macro.rs:60]: agree, removing

File: graphql/src/values/coercion.rs [hunk 4]
```diff
@@ -46,6 +46,10 @@ impl MaybeCoercible<ScalarType> for q::Value {
                 let n = num.as_i64().ok_or_else(|| q::Value::Int(num.clone()))?;
                 Ok(r::Value::Int(n))
             }
+            ("Timestamp", q::Value::Int(num)) => {
+                let n = num.as_i64().ok_or_else(|| q::Value::Int(num.clone()))?;
+                Ok(r::Value::Int(n))
+            }
```
[the remark this answers is not in the document]
Comment (dotansimha) [reply graphql/src/values/coercion.rs:52]: I fully agree. Working on a fix

File: store/postgres/src/relational_queries.rs (hunk 1 above)
Comment (dotansimha) [reply store/postgres/src/relational_queries.rs:372]: Adding

**Comment by dotansimha** [comment]
@lutter I've applied all suggestions. Tests seems to pass :) It's also using microseconds now. Thank you for the feedback!

File: graph/src/data/graphql/object_macro.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lutter) [reply graph/src/data/graphql/object_macro.rs]: This was still in the PR - I just took it out

**Review by lutter — APPROVED** [review]
Yay! I am letting this do one more CI run and will the merge it

**Pull request merged**